src/redis-benchmark.c

```diff
@@ -577,7 +594,7 @@ static void writeHandler(aeEventLoop *el, int fd, void *privdata, int mask) {
     if (c->written == 0) {
         /* Enforce upper bound to number of requests. */
         int requests_issued = 0;
-        atomicGetIncr(config.requests_issued, requests_issued, 1);
+        atomicGetIncr(config.requests_issued, requests_issued, config.pipeline);
```

Comment:
@filipecosta90 are you sure that's right? every write carries this count of commands?
i'm not familiar with the redis_benchmark code too good. doesn't it use a streaming pipeline? i.e. one that initially sends N commands, and then as soon as getting the response for one, it issues another one, so that at any given moment there are N commands in flight?

even f not, i see that we have a `c->pending` variable, maybe that's what we should be using? (rather than assuming each write has N commands)

please look into it and tell me what you found.

Reply:
> even f not, i see that we have a `c->pending` variable, maybe that's what we should be using? (rather than assuming each write has N commands)

@oranagra looking at [src/redis-benchmark.c#L716](https://github.com/redis/redis/blob/unstable/src/redis-benchmark.c#L716) we can check that c->obuf has always N pipeline commands. 

Regarding using c->pending istead of config.pipeline I believe we might get into an issue due to the initial setup commands `c->pending = config.pipeline+c->prefix_pending;` . Apart from it, and given that we only enter this condition when `c->written == 0` the value of c->pending inside that condition will always be `c->pending = config.pipeline`;

Reply:
ok...
so that that i'm sure we're on the same line.
it creates a client and an obuf with a single set of pipeline commands.
then writes to the socket, disables the write handler, enables the read handler.
when it's done reading responses it "re-sets" the client, enables the write handler and run another iteration.
so working in bursts rather than maintaining a constant in-flight pipeline.

so now i see that `clientDone` (which is called by the read handler) does this logic too:
```
    if (requests_finished >= config.requests) {
        freeClient(c);
```
so i wonder why it is needed in the write handler.
in theory it's wrong for the write handler to free the client (if it writes anything we wanna wait for the responses), but i suppose that's just dead code (the read handler will free the client rather than re-set it).

can you please look into that?

Reply:
> so now i see that clientDone (which is called by the read handler) does this logic too:

>    if (requests_finished >= config.requests) {
        freeClient(c); 

> so i wonder why it is needed in the write handler.
in theory it's wrong for the write handler to free the client (if it writes anything we wanna wait for the responses), but i suppose that's just dead code (the read handler will free the client rather than re-set it).

> can you please look into that?

@oranagra agree. It's wrong to free the client in the write handler given the chance of the replies still to be processed not being processed at all. Confirmed that [removing the freeClient within the writeHandler](https://github.com/redis/redis/pull/7947/files#diff-f37c3eacb5f6ac27e8fa820f102e8b57f0f33c7b46e5ef2ba46b8ad8a56c86a6L582) does not affect the tests and redis-benchmark still works as expected.